libraries/networking/src/udt/SendQueue.cpp

```diff
@@ -63,6 +64,44 @@ void SendQueue::queuePacket(std::unique_ptr<Packet> packet) {
     }
 }
 
+void SendQueue::queuePacketList(std::unique_ptr<PacketList> packetList) {
+    Q_ASSERT(packetList->_packets.size() > 0);
+
+    {
+        QWriteLocker locker(&_packetsLock);
+
+        auto messageNumber = getNextMessageNumber();
+
+        if (packetList->_packets.size() == 1) {
+            auto packet = packetList->takeFront<Packet>();
+            packet->setPacketPosition(Packet::PacketPosition::ONLY);
+
+            packet->writeMessageNumber(messageNumber);
+            _packets.push_back(std::move(packet));
+        } else {
+            bool haveMarkedFirstPacket = false;
+            while (!packetList->_packets.empty()) {
+                auto packet = packetList->takeFront<Packet>();
+                if (!haveMarkedFirstPacket) {
+                    packet->setPacketPosition(Packet::PacketPosition::FIRST);
+                    haveMarkedFirstPacket = true;
+                } else if (packetList->_packets.empty()) {
+                    packet->setPacketPosition(Packet::PacketPosition::LAST);
+                } else {
+                    packet->setPacketPosition(Packet::PacketPosition::MIDDLE);
+                }
+
+                packet->writeMessageNumber(messageNumber);
+
+                _packets.push_back(std::move(packet));
+            }
+        }
+    }
```

Comment:
I think a better approach here would be to do everything we need to the packet but leave them in the list and them splice the 2 lists.
It'll be faster and we'll lock for a much smaller timer.